inline_types.go

```diff
@@ -0,0 +1,690 @@
+	// from the other fields and used automatically.
+	ID string `json:"id",hash:"ignore"`
+
+	// A valid URL for the audio file.
+	AudioURL string `json:"audio_url"`
+
+	// Title.
+	Title string `json:"title"`
+
+	// Optional. Performer.
+	Performer string `json:"performer,omitempty"`
+
+	// Optional. Audio duration in seconds.
+	Duration int `json:"audio_duration,omitempty"`
+
+	// Optional. Inline keyboard attached to the message.
+	ReplyMarkup InlineKeyboardMarkup `json:"reply_markup,omitempty"`
+
+	// Optional. Content of the message to be sent instead of the audio.
+	InputMessageContent InputMessageContent `json:"input_message_content,omitempty"`
+}
+
+// Used to avoid endless recursion in MarshalJSON.
+type wrappedInlineQueryResultAudio InlineQueryResultAudio
+
+func (r *InlineQueryResultAudio) MarshalJSON() ([]byte, error) {
+	id, err := r.id()
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(struct {
+		wrappedInlineQueryResultAudio
+		Type string `json:"type"`
+		ID   string `json:"id",hash:"ignore"`
+	}{
+		wrappedInlineQueryResultAudio: wrappedInlineQueryResultAudio(*r),
+		ID:   id,
+		Type: "audio",
+	})
+}
+
+func (r *InlineQueryResultAudio) id() (string, error) {
+	if r.ID == "" {
+		return hashInlineQueryResult(r)
+	} else {
+		return r.ID, nil
+	}
+}
+
+// InlineQueryResultContact represents a contact with a phone number.
+// See also: https://core.telegram.org/bots/api#inlinequeryresultcontact
+type InlineQueryResultContact struct {
+	// Unique identifier for this result, 1-64 Bytes.
+	// If left unspecified, a 64-bit FNV-1 hash will be calculated
+	// from the other fields and used automatically.
+	ID string `json:"id",hash:"ignore"`
+
+	// Contact's phone number.
+	PhoneNumber string `json:"phone_number"`
+
+	// Contact's first name.
+	FirstName string `json:"first_name"`
+
+	// Optional. Contact's last name.
+	LastName string `json:"last_name,omitempty"`
+
+	// Optional. Inline keyboard attached to the message.
+	ReplyMarkup InlineKeyboardMarkup `json:"reply_markup,omitempty"`
+
+	// Optional. Content of the message to be sent instead of the audio.
+	InputMessageContent InputMessageContent `json:"input_message_content,omitempty"`
+
+	// Optional. Url of the thumbnail for the result.
+	ThumbURL string `json:"thumb_url,omitempty"`
+
+	// Optional. Thumbnail width.
+	ThumbWidth int `json:"thumb_width,omitempty"`
+
+	// Optional. Thumbnail height.
+	ThumbHeight int `json:"thumb_height,omitempty"`
+}
+
+// Used to avoid endless recursion in MarshalJSON.
+type wrappedInlineQueryResultContact InlineQueryResultContact
+
+func (r *InlineQueryResultContact) MarshalJSON() ([]byte, error) {
+	id, err := r.id()
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(struct {
+		wrappedInlineQueryResultContact
+		Type string `json:"type"`
+		ID   string `json:"id",hash:"ignore"`
+	}{
+		wrappedInlineQueryResultContact: wrappedInlineQueryResultContact(*r),
+		ID:   id,
+		Type: "contact",
+	})
```

Comment:
As you've just defined the struct and you're giving a value to all the struct's fields, wouldn't it be better to omit the field names? Like this:

```
	return json.Marshal(struct {
		wrappedInlineQueryResultArticle
		Type string `json:"type"`
		ID   string `json:"id",hash:"ignore"`
	}{wrappedInlineQueryResultArticle(*r), id, "article"})
```

This comment also applies to all the other inline types.